### PR TITLE
Use antlr4 for FQN instead of regex [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by amiorin**
### Describe your changes :

I worked on replacing regex for FQN with antlr4 because grammars are easier to understand than regex.

- [x] Java code
- [x] Python code
- [x] JavaScript code

# Update
I will split the PR into multiple PRs.

- [x] Java Antlr with the current logic #4066 
- [ ] JavaScript Antlr without generated code committed like the JSON schema files #4099
- [ ] Python Antlr without generated code committed like the JSON schema files
- [ ] Java Antlr with the quote as allowed char in entityName

@harshach will think about the special case of `entityName` containing `/`  and/or `"` inside `EntityLinks` and come back to me if we need to escape the `/` and/or `"`.
Redshift Table names can be anything, including `/` and `"`.
https://docs.aws.amazon.com/redshift/latest/dg/r_names.html

Nice to know
- [ ] SQLAlchemy metadata API with Redshift Table name `"`. Does it return `""` or `"`?

#
### Type of change :

- [x] Improvement

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/TableRepository.java [hunk 1]
```diff
@@ -1099,7 +1100,7 @@ private void updateColumns(
         updateColumnDataLength(stored, updated);
         updateTags(
             stored.getFullyQualifiedName(),
-            EntityUtil.getFieldName(fieldName, updated.getName(), "tags"),
+            EntityUtil.getFieldName(fieldName, quoteName(updated.getName()), "tags"),
```
Comment (sureshms) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/TableRepository.java:1103]: Please use FullyQualifiedName.build() method and make sure you are not spreading the logic of building this all over the place.

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/databases/TableResourceTest.java [hunk 2]
```diff
@@ -1446,7 +1446,7 @@ void patch_tableColumns_200_ok(TestInfo test) throws IOException {
     List<Column> columns = new ArrayList<>();
     columns.add(getColumn("c1", INT, USER_ADDRESS_TAG_LABEL).withDescription(null));
     columns.add(getColumn("c2", BIGINT, USER_ADDRESS_TAG_LABEL));
-    columns.add(getColumn("\"c.3\"", FLOAT, GLOSSARY1_TERM1_LABEL));
+    columns.add(getColumn("c.3", FLOAT, GLOSSARY1_TERM1_LABEL));
```
Comment (sureshms) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/resources/databases/TableResourceTest.java:1450]: Why are you changing this? This is a test for receiving quoted names in the API that we should unquote and store and still build the right FQN.

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/util/FullyQualifiedNameTest.java [hunk 3]
```diff
@@ -20,11 +19,7 @@ public void validate(String[] actualParts, String actualFQN) {
       assertEquals(fqn, actualFQN);
       assertEquals(parts.length, actualParts.length);
       for (int i = 0; i < parts.length; i++) {
-        if (parts[i].contains(".")) {
-          assertEquals(FullyQualifiedName.quoteName(parts[i]), actualParts[i]);
-        } else {
-          assertEquals(parts[i], actualParts[i]);
-        }
+        assertEquals(parts[i], actualParts[i]);
       }
```
Comment (sureshms) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/util/FullyQualifiedNameTest.java:23]: Why is this change being made?

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/util/FullyQualifiedNameTest.java [hunk 4]
```diff
@@ -50,21 +47,11 @@ void test_build_split() {
...
-        () -> FullyQualifiedName.quoteName("\"a")); // Error when ending quote is missing
-
-    assertThrows(
-        IllegalArgumentException.class,
-        () -> FullyQualifiedName.quoteName("a\"")); // Error when beginning quote is missing
-
-    assertThrows(
-        IllegalArgumentException.class,
-        () -> FullyQualifiedName.quoteName("a\"b")); // Error when invalid quote is present in the middle of the string
+    assertEquals("\"a\"\"b\"", FullyQualifiedName.quoteName("a\"b")); // Add quotes and quote the quote
```
Comment (sureshms) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/util/FullyQualifiedNameTest.java:53]: Why is this change being made?

File: catalog-rest-service/src/test/java/org/openmetadata/catalog/util/FullyQualifiedNameTest.java [hunk 5]
```diff
@@ -50,21 +47,11 @@ void test_build_split() {
...
+    assertEquals("\"\"\"a\"\"\"", FullyQualifiedName.quoteName("\"a\"")); // Add quotes and quote the quote
+    assertEquals("\"\"\"a.b\"\"\"", FullyQualifiedName.quoteName("\"a.b\"")); // Add quotes and quote the quote
```
Comment (sureshms) [inline catalog-rest-service/src/test/java/org/openmetadata/catalog/util/FullyQualifiedNameTest.java:55]: What is quote the quote mean?

File: openmetadata-ui/src/main/resources/ui/src/antlr/FqnListener.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+// Generated from /Users/amiorin/code/OpenMetadata/catalog-rest-service/src/main/antlr4/org/openmetadata/catalog/Fqn.g4 by ANTLR 4.9.3
+// jshint ignore: start
```
Comment (sureshms) [inline openmetadata-ui/src/main/resources/ui/src/antlr/FqnListener.js:2]: Are we going to check in generated files? Why?

**Review by sureshms — CHANGES_REQUESTED** [review]

**Comment by sureshms** [comment]
Can we make the existing functionality work based on antlr before adding new functionality? All the tests must pass without being changed. No new functionality should be added. If we additional changes are needed (I don't understand quoting the quote), we can discuss and do it later.

**Comment by amiorin** [comment]
I will split the PR into multiple PRs.

- [ ] Java Antlr with the current logic
- [ ] Java Antlr with the quote as allowed char in entityName
- [ ] Python Antlr without generated code committed like the JSON schema files
- [ ] JavaScript Antlr without generated code committed like the JSON schema files

@harshach will think about the special case of `entityName` containing `/`  and/or `"` inside `EntityLinks` and come back to me if we need to escape the `/` and/or `"`.
Redshift Table names can be anything, including `/` and `"`.
https://docs.aws.amazon.com/redshift/latest/dg/r_names.html

Nice to know
- [ ] SQLAlchemy metadata API with Redshift Table name `"`. Does it return `""` or `"`?

**Pull request closed without merging**